tutorials/scripting/c_sharp/c_sharp_signals.rst

```diff
@@ -0,0 +1,147 @@
+.. _doc_c_sharp_signals:
+
+C# Signals
+==========
+
+For a detailed explanation of signals in general, see the :ref:`doc_signals` section in the step
+by step tutorial.
+
+If it is still possible to use signals through the ``Connect``/``Disconnect`` API, C# gives us
+a more idiomatic way into the observer pattern.
+
+Signals as C# events
+--------------------
+
+To provide more type-safety, Godot signals are also all available through events. You can handle
+these events, as any other event, with the ``+=`` and ``-=`` operators.
+
+.. code-block:: csharp
+
+    Timer myTimer = GetNode<Timer>("Timer");
+    myTimer.Timeout += () => GD.Print("Timeout!");
+
+In addition, you can always access signal names associated with a node type through its nested
+``SignalName`` class. This is useful when, for example, you want to await on a signal (see :ref:`doc_c_sharp_differences_yield`).
+
+.. code-block:: csharp
+
+    await ToSignal(GetTree(), SceneTree.SignalName.ProcessFrame);
+
+.. note::
+
+    Godot will take care of disconnecting all the signals you connected through events when your
+    nodes are freed. Meaning that: as you don't need to call ``Disconnect`` on all signals you used
+    ``Connect`` on, you don't need to ``-=`` all the signals you used ``+=`` on.
+
+Custom signals as C# events
+---------------------------
+
+To declare a custom event in your C# script, use the ``[Signal]`` attribute on a delegate type.
+Note that the name of this delegate needs to end with ``EventHandler``.
+
+.. code-block:: csharp
+
+    [Signal]
+    public delegate void SayHelloEventHandler();
```

Comment:
I don't think it's noted anywhere but signal names should generally use past tense. Consider `SaidHelloEventHandler` or `GreetedEventHandler`. Or we could just use `MySignalEventHandler` like in the example that was in the C# features page.

.NET guidelines recommend using past tense for events raised _after_ the fact (e.g.: `Closed`) and present tense for events raised _before_ the fact (e.g.: `Closing`).